File: 02/js/LinkedQueue.js

class Node {
	constructor(elm, next) {
		this.elm = elm === undefined ? null : elm;
		this.next = next === undefined ? null : next;
	}

	toString() {
		return String(this._elm);
	}
}

class LinkedQueue {
	constructor() {
		this._head = null;
		this._tail = null;
		this._size = 0;
	}

	error(message) { return new Error(message) }
	getSize() { return this._size; }
	isEmpty() { return this._size === 0; }

	enqueue(elm) {
		// 初始为空时
		if( !this._tail ) {
			this._tail = new Node(elm);
			this._head = this._tail;
		} else {
			const node = new Node(elm);
			this._tail.next = node;
			this._tail = node;
		}

		this._size ++;
	}
	dequeue() {
		if( !this._head ) throw this.error('Cannot dequeue from an empty queue.');

		const delNode = this._head;
		this._head = delNode.next;
		delNode.next = null;

		if( !this._head ) this._tail = null;

		this._size --;
		return delNode.elm;
	}
	getFront() {
		if( !this._head ) throw this.error('Queue is empty.');
		return this.head.elm;
	}

	toString() {
		let res = 'front ';
		for( let cur = this._head; cur != null; cur = cur.next ) {
			res += `${cur.elm} => `;
		}
		return `${res}NULL tail`;
	}
}

module.exports = LinkedQueue;

File: 02/java/LinkedQueue.java
// 尾指针链表队列
public class LinkedQueue<E> implements Queue<E> {
	private class Node {
		public E e;
		public Node next;

		public Node(E e, Node next) {
			this.e = e;
			this.next = next;
		}

		public Node(E e) { this(e, null); }

		public Node() { this(null, null); }

		@Override
		public String toString() {
			return e.toString();
		}
	}

	private Node head, tail;
	private int size;

	public LinkedQueue() {
		head = null;
		tail = null;
		size = 0;
	}

	@Override
	public E getFront() {
		if( head == null ) {
			throw new IllegalArgumentException("Queue is empty.");
		}

		return head.e;
	}

	@Override
	public int getSize() { return size; }

	@Override
	public boolean isEmpty() { return size == 0; }

	@Override
	public void enqueue(E e) {
		if( tail == null ) {
			tail = new Node(e);
			head = tail;
		} else {
			tail.next = new Node(e);
			tail = tail.next;
		}

		size ++;
	}

	@Override
	public E dequeue() {
		if( isEmpty() ) {
			throw new IllegalArgumentException("Cannot dequeue from an empty queue.");
		}

		Node delNode = head;
		head = head.next;
		delNode.next = null;

		if( head == null ) {
			tail = null;
		}

		size --;
		return delNode.e;
	}

	@Override
	public String toString() {
		StringBuilder res = new StringBuilder();
		res.append("Queue: front ");
		for( Node cur = head.next; cur != null; cur = cur.next ) {
			res.append(cur + " => ");
		}
		res.append("NULL tail");

		return res.toString();
	}
}